File: src/lib/tours/routeVisibility.ts
/**
 * Sichtbarkeit gespeicherter Touren.
 * Default privat. Kein stilles GPS in Explore/Community/Near-me.
 */

import type { RouteVisibility, SavedRoute } from "@/types/route";
import { catalogTourIdOf } from "@/lib/tours/tourAkte";

export type { RouteVisibility };

/** Filter in Mappe / Discover-Listen. */
export type VisibilityScope = "all_mine" | "private" | "shared";

export function visibilityOf(
  route: Pick<SavedRoute, "visibility"> | { visibility?: string | null }
): RouteVisibility {
  return route.visibility === "shared" ? "shared" : "private";
}

export function isShared(
  route: Pick<SavedRoute, "visibility"> | { visibility?: string | null }
): boolean {
  return visibilityOf(route) === "shared";
}

/** Stimmen: Katalog (schon öffentlich) oder eigene Freigabe. */
export function allowsStimmen(
  route: Pick<SavedRoute, "id" | "catalogTourId" | "visibility">
): boolean {
  return stimmenTourIdOf(route) != null;
}

export function stimmenTourIdOf(
  route: Pick<SavedRoute, "id" | "catalogTourId" | "visibility">
): string | null {
  const catalog = catalogTourIdOf(route);
  if (catalog) return catalog;
  if (isShared(route)) return route.id;
  return null;
}

/** Öffentliche Explore-/Community-Listen: nur freigegebene Kopien. */
export function visibleInPublicExplore(
  route: Pick<SavedRoute, "visibility">
): boolean {
  return isShared(route);
}

/** Heatmap-Beitrag aus einer gespeicherten Tour-Geometrie: nur nach Freigabe. */
export function mayContributeSavedGeometry(
  route: Pick<SavedRoute, "visibility">
): boolean {
  return isShared(route);
}

/**
 * Ride-GPS in die Heatmap: Freeride ja; private GPX-Akte nein;
 * Katalog-Join oder Freigabe ja.
 */
export function mayContributeRideTrack(
  saved: Pick<SavedRoute, "id" | "catalogTourId" | "visibility"> | undefined
): boolean {
  if (!saved) return true;
  if (isShared(saved)) return true;
  return Boolean(catalogTourIdOf(saved));
}

export function filterSavedByVisibility<T extends Pick<SavedRoute, "visibility">>(
  routes: T[],
  scope: VisibilityScope
): T[] {
  if (scope === "all_mine") return routes;
  if (scope === "shared") return routes.filter(isShared);
  return routes.filter((r) => !isShared(r));
}

/** Sammlung-Share: nur freigegebene oder Katalog-IDs (kein privates GPS). */
export function shareableRouteIds(
  routeIds: string[],
  saved: SavedRoute[]
): string[] {
  return routeIds.filter((id) => {
    const r = saved.find((x) => x.id === id);
    if (!r) return false;
    return isShared(r) || Boolean(catalogTourIdOf(r));
  });
}

export function visibilityLabel(v: RouteVisibility): string {
  return v === "shared" ? "Freigegeben" : "Privat";
}

export function shareHonesty(route: SavedRoute): string {
  const catalog = catalogTourIdOf(route);
  const hasTrack = Boolean(route.geometry?.coordinates && route.geometry.coordinates.length >= 2);
  if (catalog && !hasTrack) {
    return "Katalog-Tour ist schon öffentlich. Freigeben macht deine Akte teilbar — der Link zeigt Name und Stats, keinen privaten Extra-Track.";
  }
  if (hasTrack) {
    return "Freigeben erzeugt einen Link. Der Link enthält eine vereinfachte Spur (Koordinaten), nicht nur den Namen. Zurück auf Privat nimmt die Tour aus Filtern und speichert den Widerruf auf dem Server, wenn du eingeloggt bist. Ohne Login gilt er nur in diesem Browser.";
  }
  return "Freigeben erzeugt einen Link mit Name und Stats — ohne Track, weil keiner gespeichert ist.";
}

File: mobile/lib/domain/tours/route_visibility.dart
import '../saved_route.dart';
import '../saved_route_note.dart';
import 'tour_akte.dart';
import '../routing/tour_filters.dart';

/// Sichtbarkeit gespeicherter Touren. Default privat. Kein stilles GPS.
abstract final class RouteVisibility {
  static const private = 'private';
  static const shared = 'shared';

  static String visibilityOf(SavedRouteMeta? meta) {
    return meta?.visibility == shared ? shared : private;
  }

  static bool isShared(SavedRouteMeta? meta) => visibilityOf(meta) == shared;

  static String? stimmenTourIdOf(String routeId, [SavedRouteMeta? meta]) {
    final m = meta ?? SavedRouteMeta.empty;
    final catalog = catalogTourIdOf(routeId, m);
    if (catalog != null) return catalog;
    if (isShared(m)) return routeId;
    return null;
  }

  static bool allowsStimmen(String routeId, [SavedRouteMeta? meta]) {
    return stimmenTourIdOf(routeId, meta) != null;
  }

  static bool visibleInPublicExplore(SavedRouteMeta? meta) => isShared(meta);

  static bool mayContributeSavedGeometry(SavedRouteMeta? meta) =>
      isShared(meta);

  /// Freeride ja; private GPX-Akte nein; Katalog-Join oder Freigabe ja.
  static bool mayContributeRide(String? routeId, SavedRouteMeta? meta) {
    if (routeId == null || routeId.trim().isEmpty) return true;
    if (isShared(meta)) return true;
    return catalogTourIdOf(routeId, meta ?? SavedRouteMeta.empty) != null;
  }

  static List<SavedRouteEntry> filter(
    List<SavedRouteEntry> routes,
    TourVisibilityKey scope,
    Map<String, SavedRouteMeta> metas,
  ) {
    if (scope == TourVisibilityKey.allMine) return routes;
    return [
      for (final r in routes)
        if (TourFilters.visibilityMatches(visibilityOf(metas[r.id]), scope)) r,
    ];
  }

  static List<String> shareableRouteIds(
    List<String> routeIds,
    Map<String, SavedRouteMeta> metas,
  ) {
    return [
      for (final id in routeIds)
        if (isShared(metas[id]) ||
            catalogTourIdOf(id, metas[id] ?? SavedRouteMeta.empty) != null)
          id,
    ];
  }

  static String shareHonesty({
    required String routeId,
    required bool hasTrack,
    SavedRouteMeta? meta,
  }) {
    final catalog = catalogTourIdOf(routeId, meta ?? SavedRouteMeta.empty);
    if (catalog != null && !hasTrack) {
      return 'Katalog-Tour ist schon freigegeben. Freigeben macht deine Tour '
          'teilbar — der Link zeigt Name und Stats, keinen privaten Extra-Track.';
    }
    if (hasTrack) {
      return 'Freigeben erzeugt einen Link. Der Link enthält eine vereinfachte '
          'Spur (Koordinaten), nicht nur den Namen. Zurück auf Privat nimmt die '
          'Tour aus Filtern und speichert den Widerruf auf dem Server, wenn du '
          'eingeloggt bist. Ohne Login gilt er nur auf diesem Gerät.';
    }
    return 'Freigeben erzeugt einen Link mit Name und Stats — ohne Track, '
        'weil keiner gespeichert ist.';
  }
}
